Libraries/LibJS/Runtime/Value.cpp

```diff
@@ -305,55 +395,71 @@ Value less_than_equals(Interpreter& interpreter, Value lhs, Value rhs)
 {
-    bool lhs_invalid = lhs.is_undefined() || lhs.is_null() || lhs.is_nan() || lhs.is_infinity();
-    bool rhs_invalid = rhs.is_undefined() || rhs.is_null() || rhs.is_nan() || rhs.is_infinity();
-
-    if (lhs_invalid && rhs_invalid)
-        return Value(0);
-
-    if (lhs_invalid || rhs_invalid)
-        return lhs_invalid ? rhs.to_number(interpreter) : lhs.to_number(interpreter);
-
-    if (!rhs.is_number() && !lhs.is_number())
-        return Value(0);
-
-    auto lhs_number = lhs.to_number(interpreter);
+    auto lhs_numeric = lhs.to_numeric(interpreter);
     if (interpreter.exception())
         return {};
-    auto rhs_number = rhs.to_number(interpreter);
+    auto rhs_numeric = rhs.to_numeric(interpreter);
     if (interpreter.exception())
         return {};
-    return Value((i32)lhs_number.as_double() | (i32)rhs_number.as_double());
+    if (both_number(lhs_numeric, rhs_numeric)) {
+        if (!lhs_numeric.is_finite_number() && !rhs_numeric.is_finite_number())
+            return Value(0);
+        if (!lhs_numeric.is_finite_number())
+            return rhs_numeric;
+        if (!rhs_numeric.is_finite_number())
+            return lhs_numeric;
+        return Value((i32)lhs_numeric.as_double() | (i32)rhs_numeric.as_double());
+    }
+    if (both_bigint(lhs_numeric, rhs_numeric))
+        return js_bigint(interpreter, lhs_numeric.as_bigint().big_integer().bitwise_or(rhs_numeric.as_bigint().big_integer()));
+    interpreter.throw_exception<TypeError>("Can't use bitwise OR operator with BigInt and other type");
+    return {};
 }
 
 Value bitwise_xor(Interpreter& interpreter, Value lhs, Value rhs)
 {
-    auto lhs_number = lhs.to_number(interpreter);
+    auto lhs_numeric = lhs.to_numeric(interpreter);
     if (interpreter.exception())
         return {};
-    auto rhs_number = rhs.to_number(interpreter);
+    auto rhs_numeric = rhs.to_numeric(interpreter);
     if (interpreter.exception())
         return {};
-    return Value((i32)lhs_number.as_double() ^ (i32)rhs_number.as_double());
+    if (both_number(lhs_numeric, rhs_numeric))
+        return Value((i32)lhs_numeric.as_double() ^ (i32)rhs_numeric.as_double());
+    if (both_bigint(lhs_numeric, rhs_numeric))
+        return js_bigint(interpreter, lhs_numeric.as_bigint().big_integer().bitwise_xor(rhs_numeric.as_bigint().big_integer()));
+    interpreter.throw_exception<TypeError>("Can't use bitwise XOR operator with BigInt and other type");
+    return {};
 }
 
 Value bitwise_not(Interpreter& interpreter, Value lhs)
 {
-    auto lhs_number = lhs.to_number(interpreter);
+    auto lhs_numeric = lhs.to_numeric(interpreter);
     if (interpreter.exception())
         return {};
-    return Value(~(i32)lhs_number.as_double());
+    if (lhs_numeric.is_number())
+        return Value(~(i32)lhs_numeric.as_double());
+    TODO();
+    // FIXME: this gives weird results:
+    // ~1 === -2
+    // ~1n === -4294967294n // ???
```

Comment:
This is a little mistake on my part, not'ing the bigint probably shouldn't not the most significant word, i.e.
(assuming 16 bit words) `0x0001 | 0x0001` should become `0x0000 | 0x1110` (the leading zeros shouldn't be negated)

but even then, we're doing sign-magnitude and even though the spec doesn't specify, everyone expects 2's complement.
while that's rather simple to do for simple arithmetic, I don't want to deal with its edge-cases in division.
This will fix it to say `~1n == -0n`, and `~~1n == 1n`, which is expected of SM.

```diff
diff --git a/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp b/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp
index 607efdd8a..bfc141f38 100644
--- a/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp
+++ b/Libraries/LibCrypto/BigInt/UnsignedBigInteger.cpp
@@ -505,10 +505,22 @@ FLATTEN void UnsignedBigInteger::bitwise_not_without_allocation(
         output.invalidate();
         return;
     }
+    if (right.length() == 0) {
+        output.set_to_0();
+        return;
+    }
+
     output.m_words.resize_and_keep_capacity(right.length());

-    for (size_t i = 0; i < right.length(); ++i)
-        output.m_words[i] = ~right.words()[i];
+    if (right.length() > 1) {
+        for (size_t i = 0; i < right.length() - 1; ++i)
+            output.m_words[i] = ~right.words()[i];
+    }
+
+    auto last_word_index = right.length() - 1;
+    auto last_word = right.words()[last_word_index];
+
+    output.m_words[last_word_index] = ((u32)0xffffffffffffffff >> __builtin_clz(last_word)) & ~last_word;
 }

 /**
diff --git a/Libraries/LibJS/Runtime/Value.cpp b/Libraries/LibJS/Runtime/Value.cpp
index 6d1d520b9..8426eb6d7 100644
--- a/Libraries/LibJS/Runtime/Value.cpp
+++ b/Libraries/LibJS/Runtime/Value.cpp
@@ -455,11 +455,7 @@ Value bitwise_not(Interpreter& interpreter, Value lhs)
         return {};
     if (lhs_numeric.is_number())
         return Value(~(i32)lhs_numeric.as_double());
-    TODO();
-    // FIXME: this gives weird results:
-    // ~1 === -2
-    // ~1n === -4294967294n // ???
-    // return js_bigint(interpreter, lhs_numeric.as_bigint().big_integer().bitwise_not());
+    return js_bigint(interpreter, lhs_numeric.as_bigint().big_integer().bitwise_not());
 }

 Value unary_plus(Interpreter& interpreter, Value lhs)
```

Reply:
but `~1n` should result in `-2n` it seems, i.e. same as `~1` results in `-2`?

Reply:
well, assuming it works in 2's complement, yes
\<basic numeric system stuff\>
1 has the lowest bit set, negating that gives us a bunch of 1's and the lsb not set, which is -1 (all bit set) - 1 (sans the firs bit):
1: `1000`
-2: `0111`
\</basic numeric system stuff\>

That's actually a decently simple system, except that division would be rather difficult to implement in terms of unsigned numbers.
unless the spec says they _must_ be in this system, we can just ignore it until it breaks something...or just disallow bitwise operations on negatives.